miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+        raise NotImplementedError()
+
+    def __len__(self):
+        return self.size()
+
+    def __neq__(self, other):
+        return not self == other
+
+
+class Struct(MemField):
+    """Dumb struct.pack/unpack field. Mainly used to factorize code.
+
+    Value is a tuple corresponding to the struct @fmt passed to the constructor.
+    """
+
+    def __init__(self, fmt):
+        self._fmt = fmt
+
+    def _pack(self, fields):
+        return struct.pack(self._fmt, *fields)
+
+    def _unpack(self, raw_str):
+        return struct.unpack(self._fmt, raw_str)
+
+    def size(self):
+        return struct.calcsize(self._fmt)
+
+    def __repr__(self):
+        return "%s(%s)" % (self.__class__.__name__, self._fmt)
+
+    def __eq__(self, other):
+        return self.__class__ == other.__class__ and self._fmt == other._fmt
+
+    def __hash__(self):
+        return hash(hash(self.__class__) + hash(self._fmt))
+
+
+class Num(Struct):
+    """Represents a number (integer or float). The number is encoded with
+    a struct-style format which must represent only one value.
+
+    TODO: use u32, i16, etc. for format.
+    """
+
+    def _pack(self, number):
+        return super(Num, self)._pack([number])
+
+    def _unpack(self, raw_str):
+        upck = super(Num, self)._unpack(raw_str)
+        if len(upck) > 1:
+            raise ValueError("Num format string unpacks to multiple values, "
+                             "should be 1")
+        return upck[0]
+
+
+class Ptr(Num):
+    """Special case of number of which value indicates the address of a
+    MemStruct. Provides deref_<field> as well as <field> when used, to set and
+    get the pointed MemStruct.
+    """
+
+    def __init__(self, fmt, dst_type, *type_args, **type_kwargs):
+        """Args:
+            fmt: (str) Num compatible format that will be the Ptr representation
+                in memory
+            dst_type: (MemStruct or MemField) the MemStruct this Ptr points to.
+                If a MemField is given, it is transformed into a MemStruct with
+                mem(TheMemField).
+            *type_args, **type_kwargs: arguments to pass to the the pointed
+                MemStruct when instanciating it (e.g. for MemStr encoding or
+                MemArray field_type).
+        """
+        if not isinstance(dst_type, MemField) and\
+                not (isinstance(dst_type, type) and\
+                        issubclass(dst_type, MemStruct)) and\
+                not dst_type == MemSelf:
+            raise ValueError("dst_type of Ptr must be a MemStruct type, a "
+                             "MemField instance, the MemSelf marker or a class "
+                             "name.")
+        super(Ptr, self).__init__(fmt)
+        if isinstance(dst_type, MemField):
+            # Patch the field to propagate the MemSelf replacement
+            dst_type._get_self_type = lambda: self._get_self_type()
```

Comment:
The two versions are actually not equivalent:
* `dst_type._get_self_type = lambda: self._get_self_type()` will make `dst_type` always call the current `_get_self_type` of `self` _at the time of the call_.
* `dst_type._get_self_type = self._get_self_type` will make `dst_type` call the `_get_self_type` function of `self` that was set _at the moment of the method assignment_. Note that `self._get_self_type` will later be patched by `MemStruct.gen_fields`.

Since inner types are created before outer types (e.g. `Ptr(_, MemSelf)` before `Ptr(_, Ptr(_, MemSelf))`), the `_get_self_type` of the outer is not patched at the moment it patches the inner class. This is why delayed evaluation of `_get_self_type` is necessary. This also is why this is a function and not an attribute.

To be honest, `MemSelf` implementation is a hacky mess; I may have been tired or confused the day I imagined that, and it stayed stuck to my mind. :)
I would really be interested in cleaner suggestion to implement that.